src/euroeval/benchmark_modules/litellm.py

```diff
@@ -344,151 +361,357 @@ def generate(self, inputs: dict) -> GenerativeModelOutput:
+                    (
+                        APIConnectionError,
+                        Timeout,
+                        ServiceUnavailableError,
+                        InternalServerError,
+                        SystemError,
+                    ),
+                ),
+                service_unavailable,
+            ),
+        ]
+
+        for predicate, handler in handlers:
+            if predicate():
+                handler()
+                return
+
+        if isinstance(error, RateLimitError):
+            raise InvalidModel(
+                f"You have encountered your rate limit for model {model_id!r}. "
+                "Skipping."
+            )
+
+        if isinstance(error, AuthenticationError):
+            raise NeedsAdditionalArgument(
+                cli_argument="--api-key",
+                script_argument="api_key=<your-api-key>",
+                run_with_cli=self.benchmark_config.run_with_cli,
+            )
+
+        raise InvalidBenchmark(
+            f"Failed to generate text. The error message was: {error}"
+        )
+
+    async def _generate_async(
+        self,
+        messages: list[dict[str, t.Any]],
+        generation_kwargs: dict[str, t.Any],
+        max_retries: int,
+        max_reruns: int,
+    ) -> tuple[list[ModelResponse], list[tuple[int, Exception]]]:
+        """Generate outputs from the model asynchronously.
+
+        Args:
+            messages:
+                The messages to pass to the model.
+            generation_kwargs:
+                The generation kwargs to pass to the model.
+            max_retries:
+                The maximum number of retries to make.
+            max_reruns:
+                The maximum number of reruns to make.
+
+        Returns:
+            A tuple containing the successful responses and the failed responses.
+        """
+        success = []
+        failures = []
+
+        to_run = list(enumerate(messages))
+
+        prev_fail_count = len(to_run)
+        rerun_count = 0
+
+        while to_run and rerun_count < max_reruns and prev_fail_count > 0:
+            if not to_run:
+                break
+
+            requests = [
+                litellm.acompletion(
+                    messages=msg, max_retries=max_retries, **generation_kwargs
+                )
+                for _, msg in to_run
+            ]
+            responses = await asyncio.gather(*requests, return_exceptions=True)
+
+            next_to_run = []
+            current_fail_count = 0
+
+            for (orig_idx, _), response in zip(to_run, responses):
+                if isinstance(response, Exception):
+                    current_fail_count += 1
+                    next_to_run.append((orig_idx, messages[orig_idx]))
+                else:
+                    success.append(response)
+
+            if current_fail_count >= prev_fail_count:
+                failures.extend(
+                    (idx, Exception("Exceeded adaptive rerun threshold"))
```

Comment:
If I understand this correctly, this exception should break the evaluation. If this is the case, then I think it'd be cleaner if we raise an `InvalidBenchmark` exception (from the `exceptions` module), as that's handled cleanly upstream, without breaking all ongoing evaluations.